core/src/main/java/org/elasticsearch/index/seqno/LocalCheckpointService.java

```diff
@@ -0,0 +1,149 @@
+
+
+    /**
+     * an ordered list of bit arrays representing pending seq nos. The list is "anchored" in {@link #firstSeqNoInProcessSeqNo}
+     * which marks the seqNo the fist bit in the first array corresponds to.
+     */
+    final LinkedList<FixedBitSet> processedSeqNo;
+    final int bitArraysSize;
+    long firstSeqNoInProcessSeqNo = 0;
+
+    /** the current local checkpoint, i.e., all seqNo lower (&lt;=) than this number have been completed */
+    volatile long checkpoint = -1;
+
+    /** the next available seqNo - used for seqNo generation */
+    volatile long nextSeqNo = 0;
+
+
+    public LocalCheckpointService(ShardId shardId, IndexSettings indexSettings) {
+        super(shardId, indexSettings);
+        bitArraysSize = indexSettings.getSettings().getAsInt(SETTINGS_BIT_ARRAYS_SIZE, DEFAULT_BIT_ARRAYS_SIZE);
+        if (bitArraysSize <= 0) {
+            throw new IllegalArgumentException("[" + SETTINGS_BIT_ARRAYS_SIZE + "] must be positive. got [" + bitArraysSize + "]");
+        }
+        processedSeqNo = new LinkedList<>();
+    }
+
+    /**
+     * issue the next sequence number
+     **/
+    public synchronized long generateSeqNo() {
+        return nextSeqNo++;
+    }
+
+    /**
+     * marks the processing of the given seqNo have been completed
+     **/
+    public synchronized void markSeqNoAsCompleted(long seqNo) {
+        // make sure we track highest seen seqNo
+        if (seqNo >= nextSeqNo) {
+            nextSeqNo = seqNo + 1;
+        }
+        if (seqNo <= checkpoint) {
+            // this is possible during recovery where we might replay an op that was also replicated
+            return;
+        }
+        FixedBitSet bitSet = getBitSetForSeqNo(seqNo);
+        int offset = seqNoToBitSetOffset(seqNo);
+        bitSet.set(offset);
+        if (seqNo == checkpoint + 1) {
+            updateCheckpoint();
+        }
+    }
+
+    /** gets the current check point */
+    public long getCheckpoint() {
+        return checkpoint;
+    }
+
+    /** gets the maximum seqno seen so far */
+    public long getMaxSeqNo() {
+        return nextSeqNo - 1;
+    }
+
+    /**
+     * moves the checkpoint to the last consecutively processed seqNo
+     * Note: this method assumes that the seqNo following the current checkpoint is processed.
+     */
+    private void updateCheckpoint() {
+        assert Thread.holdsLock(this);
+        assert checkpoint < firstSeqNoInProcessSeqNo + bitArraysSize - 1 : "checkpoint should be bellow the end of the first bit set (o.w. current bit set is completed and shouldn't be there)";
+        assert getBitSetForSeqNo(checkpoint + 1) == processedSeqNo.getFirst() : "checkpoint + 1 doesn't point to the first bit set (o.w. current bit set is completed and shouldn't be there)";
+        assert getBitSetForSeqNo(checkpoint + 1).get(seqNoToBitSetOffset(checkpoint + 1)) : "updateCheckpoint is called but the bit following the checkpoint is not set";
```

Comment:
Nit: Wrap the assert detail messages on the three preceding lines to the following lines, respectively?